### ISSUE TITLE
Could not get the Docke repos

### ISSUE THREAD

**Issue by minhnq1618** [offsecginger/CTFd-Docker-Challenges#2]
Hi, i am trying to use your plugin for a local CTF for my university but i am having trouble using it.

![image](https://user-images.githubusercontent.com/81002476/111875180-0a5a7980-89cb-11eb-96b0-b22b728b1d9d.png)

Even after i push the certs up, it still could not load the Repos, could you help me please ?

**Comment by offsecginger** [comment]
What version are you using??

**Comment by Ketherz** [comment]
I noted an issue with that, sometimes, you can be connected with TLS, but if there is no docker images on your docker the error "failed to connect to docker" is still showing. Do you have an image on the docker service already ? 
```docker image ls```

**Comment by minhnq1618** [comment]
Hi guys, thank for replying, i am using CTFd version 3.2.1.

If i am on the CTFd machine and list the images in the docker machine, i can see all the images are there.

![image](https://user-images.githubusercontent.com/81002476/111933646-771c6380-8af2-11eb-83c3-c66dc20bf8b3.png)

But on the CTFd plugin it not showing any images at all.

One thing to note that i needed to give the db.String some value for CTFd to actually load the docker plugin, does this affect how it work ?

![image](https://user-images.githubusercontent.com/81002476/111933793-c5316700-8af2-11eb-9059-08656c2bcc82.png)

**Comment by offsecginger** [comment]
Did you follow the config.json instructions?
Does the CTFd server have access to the Docker instance? Please send nmap
results from the scoreboard to the Docker server :)

**Comment by minhnq1618** [comment]
Yes, i did follow the config.json instructions

![image](https://user-images.githubusercontent.com/81002476/111934652-6e2c9180-8af4-11eb-94a1-8b6bfc935e89.png)

If i don't give the db.String any value then CTFd will give me this error :
`sqlalchemy.exc.CompileError: (in table 'docker_config', column 'ca_cert'): VARCHAR requires a length on dialect mysql`

Here is the nmap results, :/ look like i set this up wrong ? because the CTFd machine could not see the port 2376 on docker machine.

![image](https://user-images.githubusercontent.com/81002476/111934985-1b9fa500-8af5-11eb-93cf-f51cb2faa172.png)

On the docker machine i used this command to make Docker daemon listen on port 2376

`sudo dockerd --tlsverify --tlscacert=ca.pem --tlscert=server-cert.pem --tlskey=server-key.pem -H=192.168.110.5:2376 -H unix:///var/run/docker.sock`

**Comment by offsecginger** [comment]
I'm not sure about your docker settings but I can say that if your cert is
not in pem format you're going to have a bad time. It can't have the human
readable stuff just the base64 and the headers and footers.

**Comment by minhnq1618** [comment]
Hmm, all of them are in pem format right now.

![image](https://user-images.githubusercontent.com/81002476/111935392-f8c1c080-8af5-11eb-9548-24c7e468ff89.png)

![image](https://user-images.githubusercontent.com/81002476/111935428-0f681780-8af6-11eb-9188-65f032c8d4e5.png)

I am not sure what i did wrong actually :(

**Comment by offsecginger** [comment]
I would take a look at the docker config once more and ensure your server
is configured correctly to use TLS and that it's listening on the correct
port.

**Comment by Ketherz** [comment]
I ran into that kind of error, if i understood your code correctly you are using a mysql database. The ```db.String(2000)``` was a big problem for me because it was too short to contain all the value of the different files. So i used exactly this code : 

```python
class DockerConfig(db.Model):
        """
        Docker Config Model. This model stores the config for docker API connections.
        """
        id = db.Column(db.Integer, primary_key=True)
        hostname = db.Column("hostname",db.String(64), index=True)
        tls_enabled = db.Column("tls_enabled",db.Boolean,default=False, index=True)
        ca_cert = db.Column("ca_cert",db.Text)
        client_cert = db.Column("client_cert",db.Text)
        client_key = db.Column("client_key",db.Text)
        repositories = db.Column("repositories",db.String(100), index=True)```

I don't know if it's the same problem on your setup but it can be worth a try! If it's not changing anything, it's probably related to the docker port.

**Comment by minhnq1618** [comment]
Hi guys, i think you guys are right, there is something wrong with my docker port but i did everything like the document said in here :
[https://docs.docker.com/engine/security/protect-access/#use-tls-https-to-protect-the-docker-daemon-socket](https://docs.docker.com/engine/security/protect-access/#use-tls-https-to-protect-the-docker-daemon-socket)
And i still could not get any repos :/ could i ask how you guys setup the docker port and how did you guys tag the docker images ?

**Comment by Ketherz** [comment]
Hey, the only thing that i did more than what is written in the link you send is changing the ```/lib/systemd/system/docker.service``` file so that the docker daemon automatically launch itself with TLS at start...

I just replaced the Execstart parameter like that : 

```ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --tlsverify --tlscacert=path_to_key/ca.pem --tlscert=path_to_key/server-cert.pem --tlskey=path_to_key/server-key.pem -H=tcp://0.0.0.0:2375```

Maybe you should try to start with the following command ? :
``` sudo  dockerd  --tlsverify --tlscacert=ca.pem --tlscert=server-cert.pem --tlskey=server-key.pem   -H=0.0.0.0:2375```
Cause i don't really know if the -H : unix... cause the deamon to start only localy.

I'm trying things here, i don't know if it will work :'(

**Comment by offsecginger** [comment]
You should check for iptables and ufw settings. Past that I'm not sure.

**Comment by minhnq1618** [comment]
I tried everything you guys suggested and everything that i can thing of and it just not get any repo for me :/ look like CTFd just hate me or something else is missing, i will try to find time to figure out what is wrong with my setup, but now i thing i need to go for another direction for my CTF XD.

BTW, the main reason i tried to use this project is that i wanted to make the docker challenges to be able deploy for each user instances, and because i don't know how to do it manually i try to find any project on Github/Gitlab that can do it.

Do you guys know any project that can do the same or any document that you guys can share with me so i can learn to do things like that ? i really love CTFs and i want to learn more on how to develop them.

**Comment by offsecginger** [comment]
This is one of the only up to date docker plugins that works that I know
of. There was one called Whale that may not be updated. I am curious of
what is wrong in your setup. My company could likely give you a hand this
week to get you squared away, but I'm not trying for a sales pitch.

**Comment by Ketherz** [comment]
Take care, if the problem is from your docker config, you will have eventually the same problem with an other software. I know this plugin is working, so maybe it will be more interesting to try to make it work than restarting from an other out of date software. 

I just remembered an other issue that i ran on, that lead of a bad connection to the docker. It was when i generated the keys, i didn't specified the ip from where my website was hosted. This error should be logged on the website log so i assume that it's not the same error for you.

**Comment by minhnq1618** [comment]
Thank guys, our university's CTF gonna up this weekend and since there is not much time left to test and troubleshoot the issue anymore, we decided to do the old way, just have some back-up for the challenges in case anyone go rouge and messed up a challenge for everyone else. After the even i will go back and try to see how this is not working for our environment.

I actually used this project with a copy of my university's CTF and it did not work, then i went home, got myself a fresh CTFd 3.2.1 from github to my machine and tried it with the plugin and it still not work. So i am quite confuse with this XD
